Sources/Architecture/RISCV/SoftwareSingleStep.cpp

```diff
@@ -0,0 +1,171 @@
+// Copyright 2022 Saleem Abdulrasool <compnerd@compnerd.org>
+
+#include "DebugServer2/Architecture/RISCV/SoftwareSingleStep.h"
+#include "DebugServer2/Utils/Log.h"
+
+namespace ds2 {
+namespace Architecture {
+namespace RISCV {
+
+namespace {
+template <size_t Width>
+uintptr_t sext(uintptr_t value) {
+  return ((value & (1 << Width)) ? ~((value & (1 << Width)) - 1) : 0) |
+         (value & ((1 << Width + 1) - 1));
+}
+}
+
+ErrorCode PrepareSoftwareSingleStep(Target::Process *process,
+                                    BreakpointManager *manager,
+                                    CPUState const &state,
+                                    Address const &address) {
+  uintptr_t location = address.valid() ? address.value() : state.pc();
+
+  uint16_t instruction;
+  CHK(process->readMemory(location, &instruction, sizeof(instruction)));
+
+  uintptr_t destination;
+  if ((instruction & 0x3) == 0x3) {                         // RVI
+    uint32_t instruction;
+    CHK(process->readMemory(location, &instruction, sizeof(instruction)));
+
+    switch (instruction & 0x0000007f) {
+    default:
+      destination = location + 4;
+      break;
+    case 0x00000063: {                                      // BRANCH
+      uint8_t rs1 = ((instruction & 0x000f8000) >> 15);
+      uint8_t rs2 = ((instruction & 0x01f00000) >> 20);
+      uintptr_t immediate = (((instruction & 0x00000f00) >>  8) <<  1)
```

Comment:
Good question!  I am terrible with arithmetic but the compiler is good!  It can flatten this out, but in its current form it matches the specification almost verbatim.  I have less to think about this way.

<img width="539" alt="Screen Shot 2022-03-22 at 12 34 13 PM" src="https://user-images.githubusercontent.com/63311/159561322-9b5b3580-5931-4954-aa59-a57b6d25a3ec.png">